### PR TITLE
WIP attempt at parsing and evaluating spdx expressions

### PR DESCRIPTION
**Pull request by mellisio**
Thanks taking the time to contribute. This template should help guide you through the process of creating a pull request for review. Please erase any part of this template that is not relevant to your pull request:

- [ ] Have you followed the guidelines for [contributors](http://docs.libraries.io/contributorshandbook)?
- [ ] Have you checked to ensure there aren't other open pull requests on the repository for a similar change?
- [ ] Is there a corresponding ticket for your pull request?
- [ ] Have you written new tests for your changes?
- [ ] Have you successfully run the project with your changes locally?

If so then please replace this section with a link to the ticket(s) it addressed, an explanation of your change and why you think we should include it. Thanks again!

### CODE REVIEW

File: lib/parser.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+require 'treetop'
+require 'set'
+
+require File.expand_path(File.join(File.dirname(__FILE__), 'node_extensions.rb'))
```
Comment (jsonperl) [inline lib/parser.rb]: maybe `require File.expand_path("./node_extensions.rb", __FILE__)`

File: lib/parser.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+class Parser
+
+  Treetop.load(File.expand_path(File.join(File.dirname(__FILE__), 'spdx_parser.treetop')))
```
Comment (jsonperl) [inline lib/parser.rb]: maybe `Treetop.load(File.expand_path("./spdx_parser.treetop", __FILE__))`

File: lib/spdx.rb [hunk 3]
```diff
@@ -178,4 +179,23 @@ def self.special_cases # rubocop:disable Metrics/MethodLength
...
+
+  def self.valid_spdx?(spdx_string)
+    eval(Parser.parse_to_ruby(spdx_string))
+    return true
+  rescue SyntaxError, Exception
+    return false
+  end
+
+  def self.can_use_with_allow?(spdx_string, allowed_licenses)
+    licenses = allowed_licenses
```
Comment (jsonperl) [inline lib/spdx.rb]: I think we're missing some logic?

File: lib/spdx.rb [hunk 4]
```diff
@@ -178,4 +179,23 @@ def self.special_cases # rubocop:disable Metrics/MethodLength
...
+    eval(Parser.parse_to_ruby(spdx_string))
```
Comment (jsonperl) [inline lib/spdx.rb]: This is really cool, but my spidey sense tells me we're opening up an attack vector. Let's chat about it later toady.

Comment (katzj) [reply lib/spdx.rb]: Yeah, I'm super wary of `eval` here....

File: bin/update-license-files.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+curl https://raw.githubusercontent.com/spdx/license-list-data/master/json/licenses.json > ../licenses.json
```
Comment (jsonperl) [inline bin/update-license-files.sh]: Maybe add a note to the readme about running this, and add the exec flag to the file (don't know how to check that on github)

File: lib/parser.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+require "treetop"
+require "set"
+
+require_relative "spdx_grammar"
+
+class Parser
+
+  Treetop.load(File.expand_path(File.join(File.dirname(__FILE__), "spdx_parser.treetop")))
```
Comment (jsonperl) [inline lib/parser.rb]: We're skipping the pre-compilation?

**Pull request closed without merging**